design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,319 @@
+third-party project.
+
+
+### Security
+
+Having the "request" resource managed inside the API server has a better
+security profile (i.e. `UserInfo` fields are managed by the API server rather
+than a configurable webhook,
+[see](https://github.com/jetstack/cert-manager/pull/3630)).
+
+
+### cert-manager Webhook Dependency
+
+Consumers of the certificates.k8s.io `CertificateRequest` do not require
+cert-manager as a hard dependency.  cert-manager can cause issues with order of
+operation scenarios during installation or upgrades. For example, the
+`CertificateRequest` resource needs to be installed and the webhook ready for
+consumers/clients of this API to make requests.
+
+In some setups, the cert-manager installation is handled by two separate teams:
+
+- The cert-manager CRDs and webhook are installed and managed by the
+  operations team, which allows the operations team to enforce cluster
+  policies;
+- The cert-manager controller, cainjector and acmesolver are installed and
+  managed by the developer team.
+
+By supporting the built-in `CertificateSigningRequest` resource, we decrease the
+coupling between the operations team and the developer team.  Since the
+`CertificateSigningRequest` resource is built-in, it does not need any
+`CustomResourceDefinition` installation, and the webhook for this resource is
+integrated to the API server.
+
+
+### CertificateSigningRequest Migration Path
+
+Support for the `CertificateSigningRequest` resource in cert-manager is a
+stepping stone to migrating to this resource instead of the `CertificateRequest`
+resource, if the project does indeed want to migrate.
+
+
+### Goals
+
+* Add `CertificateSigningRequest` signer controllers to cert-manager for all
+    `Issuer` types
+* Pave a way for replacement of the `CertificateRequest` resource for `CertificateSigningRequest`
+
+### Non-Goals
+
+* Remove the `CertificateRequest` resource before `CertificateSigningRequest` is
+    fully supported and a migration path is in place
+
+---
+
+## Reasons for Approach
+
+Below are a summary of properties of the `CertificateSigningRequest` which
+cause issues or caveats which need to be addressed when adding support for the
+resource.
+
+### Read all access to CertificateSigningRequests
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. This means that in order for users to have the
+same debugging and usage capabilities of the `CertificateRequest`, users will
+have the ability to list all requests in the cluster, rather than being scoped to
+specific namespaces. This is a regression in security.
+
+A solution to this problem could be given to users by providing examples or
+guides to external tooling (e.g. [Open Policy Agent
+(OPA)](https://github.com/open-policy-agent/opa), whereby cluster administrators
+can limit users access to certain `CertificateSigningRequests` based on label
+selectors.
```

Comment:
This is not possible today.  Admission does not run on read requests.  And authorization has no concept of label or field selectors.  You could create an aggregated API that can give users a list view on the subset of CSRs that they created.  Such an API could even be namespace scoped and support multiple layers of filtering.

Reply:
Thanks @enj, this is an important point that passed me. This means an even larger limitation on this.